src/transports/ipc.rs

```diff
@@ -0,0 +1,376 @@
+use std::io::{self, Read, Write};
+use std::path::Path;
+
+use self::tokio_core::reactor;
+use self::tokio_core::io::{ReadHalf, WriteHalf, Io};
+use self::tokio_uds::UnixStream;
+
+use futures::{self, Sink, Stream, Future, BoxFuture};
+use futures::sync::{oneshot, mpsc};
+use helpers;
+use rpc::{self, Value as RpcValue};
+use {Transport, Error as RpcError};
+
+macro_rules! try_nb {
+  ($e:expr) => (match $e {
+    Ok(t) => t,
+    Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+      return Ok(futures::Async::NotReady)
+    }
+    Err(e) => {
+      warn!("Error: {:?}", e);
+      return Err(())
+    },
+  })
+}
+
+/// Error returned while initializing IPC transport
+#[derive(Debug)]
+pub enum IpcError {
+  /// Wrapped IO error
+  Io(io::Error),
+}
+
+impl From<io::Error> for IpcError {
+  fn from(err: io::Error) -> Self { IpcError::Io(err) }
+}
+
+type Pending = oneshot::Sender<Result<RpcValue, RpcError>>;
+
+/// Event Loop Handle.
+/// NOTE: Event loop is stopped when handle is dropped!
+pub struct EventLoopHandle {
+  thread: Option<thread::JoinHandle<()>>,
+  done: Arc<atomic::AtomicBool>,
+}
+
+impl Drop for EventLoopHandle {
+  fn drop(&mut self) {
+    self.done.store(true, atomic::Ordering::Relaxed);
+    self.thread.take()
+      .expect("We never touch thread except for drop; drop happens only once; qed")
+      .join()
+      .expect("Thread should shut down cleanly.");
+  }
+}
+
+/// Unix Domain Sockets (IPC) transport
+pub struct Ipc {
+  id: atomic::AtomicUsize,
+  write_sender: mpsc::Sender<String>,
+  read_sender: mpsc::Sender<(usize, Pending)>,
+}
+
+impl Ipc {
+  /// Create new IPC transport within existing Event Loop.
+  pub fn new<P>(path: P, handle: &reactor::Handle) -> Result<Self, IpcError> where
+    P: AsRef<Path>,
+  {
+    trace!("Connecting to: {:?}", path.as_ref());
+    let stream = UnixStream::connect(path, handle)?;
+    Self::with_stream(stream, handle)
+  }
+
+  /// Creates new IPC transport from existing `UnixStream` and `Handle`
+  fn with_stream(stream: UnixStream, handle: &reactor::Handle) -> Result<Self, IpcError> {
+    let (read, write) = stream.split();
+    let (write_sender, write_receiver) = mpsc::channel(2);
+    let (read_sender, read_receiver) = mpsc::channel(2);
+
+    let r = ReadStream {
+      read: read,
+      incoming: read_receiver,
+      pending: Default::default(),
+      buffer: vec![],
+      current_pos: 0,
+    };
+
+    let w = WriteStream {
+      write: write,
+      incoming: write_receiver,
+      state: WriteState::WaitingForRequest,
+    };
+
+    handle.spawn(r);
+    handle.spawn(w);
+
+    Ok(Ipc {
+      id: atomic::AtomicUsize::new(1),
+      write_sender: write_sender,
+      read_sender: read_sender,
+    })
+  }
+
+  /// Create new IPC transport with separate event loop.
+  /// NOTE: Dropping event loop handle will stop the transport layer!
+  pub fn with_event_loop<P>(path: P) -> Result<(EventLoopHandle, Self), IpcError> where
```

Comment:
the `with_` naming convention implies that the thing will be provided to the function (like `with_capacity`)

So to me it would make more sense for you to switch the names of `new` and `with_event_loop`